[WIP] Redfish Client & Redfish OOB Driver

GabrielBrascher opened the pull request:
Implementation overview:
- Redfish Java client: this PR adds a Java Client for Redfish; thus, making Redfish actions available to the HA workflow via an OOB driver.
- OOB driver: a new Out-of-band driver was created for Redfish, allowing to integrate the Redfish Client with the CloudStack Out-of-band management implementation.

This PR tracks and helps on monitoring and developing the Redfish OOBM implementation

**Note:** this feature was only possible with the help of our friends at [Yourhosting](https://www.yourhosting.nl/).

GabrielBrascher committed:
Redfish Rest Client

+ Redfish OOB implementation
- OOB Redfish Driver
- RedfishClient

GabrielBrascher commented:
Testing the plugin on CloudStack workflow. The following log shows that RedfishOutOfBandManagementDriver is propperly loaded into the CloudStack.
```
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loaded module context [nested-cloudstack] in 109 ms
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/redfish/spring-redfish-context.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/outofbandmanagement/spring-core-lifecycle-outofbandmanagement-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/core/spring-core-lifecycle-core-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/system/spring-framework-config-system-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/system/spring-core-system-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/system/spring-contrail-system-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/bootstrap/spring-bootstrap-context-inheritable.xml]
2020-01-08 05:41:06,588 DEBUG [o.a.c.f.c.i.ConfigDepotImpl] (main:null) (logid:) Retrieving keys from RedfishOutOfBandManagementDriver
2020-01-08 05:41:06,660 DEBUG [o.a.c.s.l.r.ExtensionRegistry] (main:null) (logid:) Registering extension [REDFISH] in [Out Of Band Management Drivers Registry]
2020-01-08 05:41:06,660 DEBUG [o.a.c.s.l.r.RegistryLifecycle] (main:null) (logid:) Registered org.apache.cloudstack.outofbandmanagement.driver.redfish.RedfishOutOfBandManagementDriver@10683f87
2020-01-08 05:41:06,660 INFO  [o.a.c.s.l.CloudStackExtendedLifeCycle] (main:null) (logid:) Configuring CloudStack Components
2020-01-08 05:41:06,661 INFO  [o.a.c.s.l.CloudStackExtendedLifeCycle] (main:null) (logid:) Done Configuring CloudStack Components
2020-01-08 05:41:06,661 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loaded module context [redfish] in 150 ms
2020-01-08 05:41:06,663 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Starting module [system]```

wido commented on plugins/outofbandmanagement-drivers/redfish/src/main/java/org/apache/cloudstack/outofbandmanagement/driver/redfish/RedfishOutOfBandManagementDriver.java:
```diff
@@ -0,0 +1,121 @@
+            true, ConfigKey.Scope.Global);
+
+    public static final ConfigKey<Boolean> USE_HTTPS = new ConfigKey<Boolean>("Advanced", Boolean.class, "redfish.use.https", "false",
+            "If set to false the redfish client will ignore SSL certificate validation when sending requests to a Redfish server.", true, ConfigKey.Scope.Global);
```
This description seems odd. I think it was a copy/paste from the previous config key?

I think this one should say something like: "Use HTTPS/SSL for all connections"

wido commented on utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:
```diff
@@ -0,0 +1,423 @@
+            if (ignoreSSL) {
+                client = ignoreSSLCertValidator();
+            } else {// TODO spawn a regular http client
+                throw new RedfishException("Redfish Client does not support SSL certificates. Please set the client to ignore SSL certificates.");
```
Why is this needed? Because you can just spawn a regular HTTP client without anything else. It will then complain if it can't validate the SSL certificate.

wido commented on utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:
```diff
@@ -0,0 +1,423 @@
+            if (ignoreSSL) {
+                client = ignoreSSLCertValidator();
+            } else {// TODO spawn a regular http client
+                throw new RedfishException("Redfish Client does not support SSL certificates. Please set the client to ignore SSL certificates.");
```
Same as with executeGetRequest.

Can't these share the same code?

wido commented on utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:
```diff
@@ -0,0 +1,423 @@
+    private String validateAddressAndPrepareForUrl(String hostAddress) {
+        if (NetUtils.isValidIp6(hostAddress)) {
+            return String.format("[%s]", hostAddress);
+        } else if (NetUtils.isValidIp6(hostAddress)) {
```
This should validate for a valid IPv4 and not IPv6 I think?

wido commented on utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:
```diff
@@ -0,0 +1,423 @@
+        } else if (NetUtils.isValidIp6(hostAddress)) {
+            return hostAddress;
+        } else {
+            throw new RedfishException(String.format("Redfish host address '%s' is not a valid IPv4 or IPv6 address", hostAddress));
```
This shouldn't be the case always. Because a hostname should also be supported.

People might have a DNS entry pointing to the iDRAC/IPMI of their server(s).

GabrielBrascher replied:
The `InternetDomainName.isValid(hostAddress)` validation should be enough to validate a domain name.

I will be soon committing a few updates,  one of them is a new JUnit test that ensures this method to accept a domain name (e.g. mydomain.org).

wido commented on utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:
```diff
@@ -0,0 +1,423 @@
+            throw new RedfishException("Failed to process system Response", e);
+        }
+
+        String open = "/redfish/v1/Systems/";
```
Re-use SYSTEMS_URL_PATH here?

wido reviewed (commented):
Left a few comments.

Looking good though!

DaanHoogland committed:
Volume migration (#4043)

* Update AncientDataMotionStrategy.java

fix When secondary storage usage is> 90%, VOLUME migration across primary storage will cause the migration to fail and lose VOLUME

* Update AncientDataMotionStrategy.java

Volume is migrated across Primary storage. If no secondary storage is available(Or used capacity> 90% ), the migration is canceled.
Before modification, if secondary storage cannot be found, copyVolumeBetweenPools return NUll

copyAsync considers answer = null to be a sign of successful task execution, so it deletes the VOLUME on the old primary storage. This is the root cause of data loss, because VOLUME did not perform the migration at all.

* code in comment removed

Co-authored-by: div8cn <35140268+div8cn@users.noreply.github.com>
Co-authored-by: Daan Hoogland <dahn@onecht.net>

Spaceman1984 committed:
Fixed guest vlan range going missing when using zone wizzard (#4042)

DaanHoogland committed:
Merge release branch 4.13 to master

* 4.13:
  Fixed guest vlan range going missing when using zone wizzard (#4042)
  Volume migration (#4043)

GabrielBrascher committed:
Enhance code, exception handlers, and fix authentication issue

GabrielBrascher commented:
Closing Draft PR.

- _Official_ PR is at https://github.com/apache/cloudstack/pull/4175
- new branch is [pcextreme:redfish-oob](https://github.com/PCextreme/cloudstack/tree/redfish-oob)
